node/src/components/consensus/protocols/highway/synchronizer.rs

```diff
@@ -349,13 +356,18 @@ impl<I: NodeIdT, C: Context + 'static> Synchronizer<I, C> {
                 }
                 // We are still missing a dependency. Store the vertex in the map and request
                 // the dependency from the sender.
-                // Find the first dependency that `pv` needs that we haven't synchronized yet
-                // and request it from the sender of `pv`. Since it relies on it, it should have
-                // it as well.
-                let transitive_dependency = self.find_transitive_dependency(dep.clone());
                 let sender = pv.sender().clone();
                 let ser_msg = HighwayMessage::RequestDependency(transitive_dependency).serialize();
-                outcomes.push(ProtocolOutcome::CreatedTargetedMessage(ser_msg, sender));
+                if !outcomes.iter().any(|o| match o {
```

Comment:
I'd prefer to either make `outcomes` a set (and then turn it into a `Vec` only at the end of the method) or, even better:
Add a set `requested_deps: HashSet<(I, Dependency<C>)>`, so we don't even serialize the same message twice.

Although neither of those will prevent us from requesting the same vertex twice from the same peer in two separate `pop_vertex_to_add` calls.

Reply:
> Although neither of those will prevent us from requesting the same vertex twice from the same peer in two separate `pop_vertex_to_add` calls.

Deduplicating in-flight requests was not the goal of this PR.

> I'd prefer to either make outcomes a set (and then turn it into a `Vec` only at the end of the method) or, even better:
Add a set `requested_deps: HashSet<(I, Dependency<C>)>`, so we don't even serialize the same message twice.

Sure, I was too lazy to do it :)

Reply:
Done in [`12bb6c2` (#1312)](https://github.com/CasperLabs/casper-node/pull/1312/commits/12bb6c28701da5269fcc8d5cec73b2509afa6f36).